### PR TITLE
#137005171 Actionable analytics

### PR DESCRIPTION
**Pull request by andela-aolaniran**
#### What does this PR do?
It implements actionable analytics option for the admin
#### Description of Task to be completed?
Data driven product presentation, and performance analysis.
#### How should this be manually tested?
As an admin, select the actionable analytics option from your menu bar to view the page showing actionable details.
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
#137005171 Actionable Analytics
#### Screenshots (if appropriate)
screen shot 2017-01-31 at 5 51 03 pm

#### Questions:
N/A

### CODE REVIEW

File: server/methods/catalog.js [hunk 1]
```diff
@@ -1269,5 +1269,24 @@ Meteor.methods({
...
+
+  /**
+   * update the number of times a product has been viewed (for analytics purposes)
+   * @param{String} productId - Id of the product views to be updated
+   * @return{Object} - Updated product
+   */
+  "products/updateViews": (handle) => {
+    check(handle, String);
+    const product = Products.findOne({handle: handle});
+    console.log(`Handle: ${handle}: ${product._id}, Views: ${product.views}`);
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply server/methods/catalog.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+      "desc"
+    );
+  },
+  products() {
+    const instance = Template.instance();
+    const productsAnalytics = instance.state.get("productsAnalytics");
+    const products = [];
+    for (key in productsAnalytics) {
+      if (key) {
+        products.push(productsAnalytics[key]);
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        return Date.parse(order.date);
+      },
+      "desc"
+    );
+  },
+  products() {
+    const instance = Template.instance();
+    const productsAnalytics = instance.state.get("productsAnalytics");
+    const products = [];
+    for (key in productsAnalytics) {
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    );
+  },
+  statements() {
+    const statements = [];
+    const instance = Template.instance();
+    const analyticsStatement = instance.state.get("analyticsStatement");
+    for (key in analyticsStatement) {
+      if (key) {
+        statements.push(analyticsStatement[key]);
+        analyticsStatement[key].totalSales = formatPriceString(analyticsStatement[key].totalSales);
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+      "desc"
+    );
+  },
+  statements() {
+    const statements = [];
+    const instance = Template.instance();
+    const analyticsStatement = instance.state.get("analyticsStatement");
+    for (key in analyticsStatement) {
+      if (key) {
+        statements.push(analyticsStatement[key]);
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        return product.salesSorter;
+      },
+      "desc"
+    );
+  },
+  statements() {
+    const statements = [];
+    const instance = Template.instance();
+    const analyticsStatement = instance.state.get("analyticsStatement");
+    for (key in analyticsStatement) {
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    );
+  },
+  topEarning() {
+    const products = [];
+    const instance = Template.instance();
+    const analytics = instance.state.get("analytics");
+    for (key in analytics) {
+      if (key) {
+        products.push({
+          product: key,
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+  },
+  bestSelling() {
+    const products = [];
+    const instance = Template.instance();
+    const analytics = instance.state.get("analytics");
+    for (key in analytics) {
+      if (key) {
+        products.push({
+          product: key,
+          quantitySold: analytics[key].quantitySold
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/index.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+//import "./templates/overview/overview.html";
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/index.js]: Fixed

File: imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+  // a Day represented in milliseconds
+  const oneDay = 1000 * 60 * 60 * 24;
+  // Calculate the difference in milliseconds
+  const difference = date2 - date1;
+  // Convert back to days and return
+  return Math.round(difference / oneDay);
+}
+
+Template.actionableAnalytics.onCreated(function () {
+  this.state = new ReactiveDict();
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/actionable-analytics/client/templates/actionableAnalytics.js]: Fixed

File: imports/plugins/core/revisions/server/publications.js [hunk 11]
```diff
@@ -9,7 +9,7 @@ import { Reaction } from "/server/api";
  */
 Meteor.publish("ProductRevisions", function (productIds) {
   check(productIds, Array);
-
+  Logger.info("ProductRevisions method called");
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/core/revisions/server/publications.js:13]: Fixed

File: imports/plugins/included/product-detail-simple/client/templates/productDetailSimple.js [hunk 12]
```diff
@@ -1,6 +1,12 @@
 import { ProductDetailContainer } from "../containers";
 import { isRevisionControlEnabled } from "/imports/plugins/core/revisions/lib/api";
 import { Template } from "meteor/templating";
+import { Reaction } from "/client/api";
+
+Template.productDetailSimple.onCreated(() => {
+  const handle = Reaction.Router.current().params.handle;
+  Meteor.call("products/updateViews", handle);
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply imports/plugins/included/product-detail-simple/client/templates/productDetailSimple.js]: Fixed

File: server/methods/catalog.js [hunk 13]
```diff
@@ -1269,5 +1269,24 @@ Meteor.methods({
 
     Logger.debug("invalid product visibility ", productId);
     throw new Meteor.Error(400, "Bad Request");
+  },
+
+  /**
```
[the remark this answers is not in the document]
Comment (andela-aolaniran) [reply server/methods/catalog.js:1274]: Fixed

**Review by olawalejarvis — APPROVED** [review]
Nice work

**Review by vonvick — APPROVED** [review]
LGTM

**Pull request merged**